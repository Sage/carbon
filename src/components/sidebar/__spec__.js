import React from 'react';
import TestUtils from 'react/lib/ReactTestUtils';
import { Sidebar } from './sidebar';
import Textbox from './../textbox';
import Icon from './../icon';
import { shallow } from 'enzyme';
import { elementsTagTest, rootTagTest } from '../../utils/helpers/test';

describe('Sidebar', () => {
  let instance, leftInstance, spy;

  beforeEach(() => {
    spy = jasmine.createSpy();

    instance = TestUtils.renderIntoDocument(
      <Sidebar
        onCancel={ spy }
        open={ true }
        className='custom-class'
      >
        <Textbox />
        <Textbox />
        <Textbox />
      </Sidebar>
    );

    leftInstance = TestUtils.renderIntoDocument(
      <Sidebar
        onCancel={ spy }
        open={ true }
        size='small'
        position='left'
        enableBackgroundUI={ true }
      />
    );
  });

  describe('componentDidUpdate', () => {
    describe('when the Sidebar is open', () => {
      beforeEach(() => {
        instance = TestUtils.renderIntoDocument(
          <Sidebar open={ true } onCancel={ spy } />
        );
      });

      it('sets up event listeners to resize and close the Sidebar', () => {
        let spy = spyOn(window, 'addEventListener');
        instance.componentDidUpdate();
        expect(spy.calls.count()).toEqual(1);
        expect(window.addEventListener).toHaveBeenCalledWith('keyup', instance.closeModal);
      });

      describe('when the Sidebar is already listening', () => {
        it('does not set up event listeners', () => {
          let spy = spyOn(window, 'addEventListener');
          instance.listening = true;
          instance.componentDidUpdate();
          expect(spy.calls.count()).toEqual(0);
          expect(window.addEventListener).not.toHaveBeenCalled();
        });
      });
    });

    describe('when the Sidebar is closed', () => {
      beforeEach(() => {
        instance = TestUtils.renderIntoDocument(
          <Sidebar onCancel={ spy } />
        );
      });

      it('removes event listeners for resize and closing', () => {
        let spy = spyOn(window, 'removeEventListener');
        instance.componentDidUpdate();
        expect(spy.calls.count()).toEqual(1);
        expect(window.removeEventListener).toHaveBeenCalledWith('keyup', instance.closeModal);
      });
    });
  });

  describe('sidebarClasses', () => {
    it('returns a base sidebar class', () => {
      expect(instance.sidebarClasses).toMatch('carbon-sidebar__sidebar carbon-sidebar__sidebar--right carbon-sidebar__sidebar--medium');
    });

    it('returns a position modifier class based on props', () => {
      expect(leftInstance.sidebarClasses).toMatch('carbon-sidebar__sidebar--left');
    });

    it('returns a size modifier class based on props', () => {
      expect(leftInstance.sidebarClasses).toMatch('carbon-sidebar__sidebar--small');
    });
  });

  describe('mainClasses', () => {
    it('sets the sidebar class', () => {
      expect(leftInstance.mainClasses).toEqual('carbon-sidebar');
    });

    it('appends any additional passed classes', () => {
      expect(instance.mainClasses).toEqual('carbon-sidebar custom-class');
    });
  });

  describe('sidebarHTML', () => {
    it('returns the rendered sidebar', () => {
      expect(TestUtils.findRenderedDOMComponentWithClass(instance, 'carbon-sidebar__sidebar')).toBeTruthy();
    });

    it('renders a close icon', () => {
      let icon = TestUtils.findRenderedComponentWithType(instance, Icon);
      expect(icon.props.type).toEqual('close');
    });

    it('renders children within the sidebar', () => {
      let textboxes = TestUtils.scryRenderedDOMComponentsWithClass(instance, 'carbon-textbox');
      expect(textboxes.length).toEqual(3);
    });
  });

  describe('render', () => {
    describe('when sidebar is open', () => {
      it('renders a parent div', () => {
        expect(TestUtils.findRenderedDOMComponentWithClass(instance, 'carbon-sidebar')).toBeTruthy();
      });

      it('renders a background div', () => {
        expect(TestUtils.findRenderedDOMComponentWithClass(instance, 'carbon-modal__background')).toBeTruthy();
      });

      it('renders a sidebar', () => {
        expect(TestUtils.findRenderedDOMComponentWithClass(instance, 'carbon-sidebar__sidebar')).toBeTruthy();
      });
    });

    describe('when sidebar is closed', () => {
      beforeEach(() => {
        instance = TestUtils.renderIntoDocument(
          <Sidebar
            onCancel={ spy }
            open={ false }
          >
          </Sidebar>
        );
      });

      it('renders a parent div', () => {
        expect(TestUtils.findRenderedDOMComponentWithClass(instance, 'carbon-sidebar')).toBeTruthy();
      });

      it('does not render a background div', () => {
        expect(TestUtils.scryRenderedDOMComponentsWithClass(instance, 'carbon-sidebar__background').length).toEqual(0);
      });

      it('does not render a sidebar', () => {
        expect(TestUtils.scryRenderedDOMComponentsWithClass(instance, 'carbon-sidebar__sidebar').length).toEqual(0);
      });
    });
  });

  describe('Behaviour', () => {
    describe('clicking the close icon sidebar', () => {
      it('closes the sidebar', () => {
        let icon = TestUtils.findRenderedDOMComponentWithClass(instance, 'carbon-sidebar__close-icon');
        TestUtils.Simulate.click(icon);
        expect(spy).toHaveBeenCalled();
      });
    });

    describe('when there is no onCancel prop', () => {
      beforeEach(() => {
        instance = TestUtils.renderIntoDocument(
          <Sidebar open={ true }>
            <Textbox />
          </Sidebar>
        );
      });

      it('does not render a close icon', () => {
        let icon = TestUtils.scryRenderedDOMComponentsWithClass(instance, 'carbon-sidebar__close-icon');
        expect(icon.length).toEqual(0);
      });
    });
  });

  describe("tags", () => {
    describe("on component", () => {
      let wrapper = shallow(
        <Sidebar
          data-element='bar'
          onCancel={ () => {} }
          onConfirm={ () => {} }
          data-role='baz'
        />
      );

      it('include correct component, element and role data tags', () => {
        rootTagTest(wrapper, 'sidebar', 'bar', 'baz');
      });
    });

    describe("on internal elements", () => {
      let wrapper = shallow(
        <Sidebar
          open={ true }
          title='Test'
          onCancel={ () => {} }
          onConfirm={ () => {} }
        />
      );

      elementsTagTest(wrapper, [
        'close'
      ]);
    });
  });
});
