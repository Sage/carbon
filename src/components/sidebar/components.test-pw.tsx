import React, { useState, useRef } from "react";

import Typography from "../../../src/components/typography";
import Button from "../button";
import Sidebar, { SidebarProps } from ".";
import Box from "../box";
import Toast from "../toast";
import Textbox from "../textbox";
import Dialog from "../dialog";

export const Default = ({
  open = true,
  restoreFocusOnClose,
}: {
  open?: boolean;
  restoreFocusOnClose?: boolean;
}) => {
  const [isOpen, setIsOpen] = useState(open);
  const onCancel = () => {
    setIsOpen(false);
  };
  return (
    <>
      <Button onClick={() => setIsOpen(true)}>Open sidebar</Button>
      <Sidebar
        aria-label="sidebar"
        open={isOpen}
        onCancel={onCancel}
        restoreFocusOnClose={restoreFocusOnClose}
      >
        <Box mb={2}>
          <Button buttonType="primary">Test</Button>
          <Button buttonType="secondary" ml={2}>
            Last
          </Button>
        </Box>
        <Box mb="3000px">Main content</Box>
      </Sidebar>
    </>
  );
};

export const DefaultNested = () => {
  const [isFirstSidebarOpen, setIsFirstSidebarOpen] = useState(false);
  const [isNestedSidebarOpen, setIsNestedSidebarOpen] = useState(false);
  return (
    <>
      <Button onClick={() => setIsFirstSidebarOpen(true)}>
        Open First Sidebar
      </Button>
      <Sidebar
        open={isFirstSidebarOpen}
        onCancel={() => setIsFirstSidebarOpen(false)}
      >
        <Button onClick={() => setIsNestedSidebarOpen(true)}>
          Open Nested Sidebar
        </Button>
        <Sidebar
          open={isNestedSidebarOpen}
          onCancel={() => setIsNestedSidebarOpen(false)}
        >
          <Box mb={2}>
            <Button buttonType="primary">Test</Button>
            <Button buttonType="secondary" ml={2}>
              Last
            </Button>
          </Box>
        </Sidebar>
      </Sidebar>
    </>
  );
};

export const SidebarComponent = (props: Partial<SidebarProps>) => {
  return (
    <>
      <Sidebar
        aria-label="sidebar"
        open
        position="right"
        size="medium"
        {...props}
      >
        <Box mb={2}>
          <Button buttonType="primary">Test</Button>
          <Button buttonType="secondary" ml={2}>
            Last
          </Button>
        </Box>
        <Box mb="3000px">Main content</Box>
      </Sidebar>
    </>
  );
};

export const SidebarComponentWithOnCancel = (props: Partial<SidebarProps>) => {
  const [isOpen, setIsOpen] = useState(true);
  const handleOnCancel = () => {
    setIsOpen(false);
  };
  return (
    <>
      <Sidebar
        aria-label="sidebar"
        open={isOpen}
        position="right"
        size="medium"
        onCancel={handleOnCancel}
        {...props}
      >
        <Box mb={2}>
          <Button buttonType="primary">Test</Button>
          <Button buttonType="secondary" ml={2}>
            Last
          </Button>
        </Box>
        <Box mb="3000px">Main content</Box>
      </Sidebar>
    </>
  );
};

export const SidebarComponentWithHeading = (props: Partial<SidebarProps>) => {
  return (
    <>
      <Sidebar
        aria-label="sidebar"
        open
        position="right"
        size="medium"
        header={<Typography variant="h3">Sidebar Header</Typography>}
        {...props}
      >
        <Box mb={2}>
          <Button buttonType="primary">Test</Button>
          <Button buttonType="secondary" ml={2}>
            Last
          </Button>
        </Box>
        <Box mb="3000px">Main content</Box>
      </Sidebar>
    </>
  );
};

export const SidebarComponentWithDarkHeading = (
  props: Partial<SidebarProps>,
) => {
  return (
    <>
      <Sidebar
        aria-label="sidebar"
        open
        position="right"
        size="medium"
        header={
          <Typography variant="h3" color="white">
            Sidebar Header
          </Typography>
        }
        headerVariant="dark"
        {...props}
      >
        <Box mb={2}>
          <Button buttonType="primary">Test</Button>
          <Button buttonType="secondary" ml={2}>
            Last
          </Button>
        </Box>
        <Box mb="3000px">Main content</Box>
      </Sidebar>
    </>
  );
};

export const SidebarComponentWithSubHeading = (
  props: Partial<SidebarProps>,
) => {
  return (
    <>
      <Sidebar
        aria-label="sidebar"
        open
        position="right"
        size="medium"
        header={<Typography variant="h3">Sidebar Header</Typography>}
        subHeader={
          <Button iconType="chevron_left_thick" buttonType="tertiary">
            Action
          </Button>
        }
        {...props}
      >
        <Box mb={2}>
          <Button buttonType="primary">Test</Button>
          <Button buttonType="secondary" ml={2}>
            Last
          </Button>
        </Box>
        <Box mb="3000px">Main content</Box>
      </Sidebar>
    </>
  );
};

export const SidebarBackgroundScrollTestComponent = () => {
  return (
    <Box height="2000px" position="relative">
      <Box
        data-element="test-box"
        height="100px"
        position="absolute"
        bottom="0px"
      >
        I should not be scrolled into view
      </Box>
      <Sidebar open onCancel={() => {}}>
        <Textbox label="textbox" />
      </Sidebar>
    </Box>
  );
};

export const SidebarBackgroundScrollWithOtherFocusableContainers = () => {
  const toast1Ref = useRef(null);
  const toast2Ref = useRef(null);
  return (
    <Box height="2000px" position="relative">
      <Box
        data-element="test-box"
        height="100px"
        position="absolute"
        bottom="0px"
      >
        I should not be scrolled into view
      </Box>
      <Sidebar
        open
        onCancel={() => {}}
        focusableContainers={[toast1Ref, toast2Ref]}
      >
        <Textbox label="textbox" />
      </Sidebar>
      <Toast open onDismiss={() => {}} ref={toast1Ref} targetPortalId="stacked">
        Toast message 1
      </Toast>
      <Toast open onDismiss={() => {}} ref={toast2Ref} targetPortalId="stacked">
        Toast message 2
      </Toast>
    </Box>
  );
};

export const SidebarComponentFocusable = (props: Partial<SidebarProps>) => {
  const [setIsDialogOpen] = React.useState(false);
  const [isToastOpen, setIsToastOpen] = React.useState(false);
  const toastRef = React.useRef(null);
  const CUSTOM_SELECTOR = "button, .focusable-container input";
  return (
    <>
      <Sidebar
        open
        onCancel={() => setIsDialogOpen}
        header={<Typography variant="h3">Sidebar header</Typography>}
        focusableContainers={[toastRef]}
        focusableSelectors={CUSTOM_SELECTOR}
        {...props}
      >
        <Box className="focusable-container">
          <Textbox label="First Name" />
        </Box>
        <Box>
          <Textbox label="Surname" />
        </Box>
        <Box className="focusable-container">
          <Button
            buttonType="primary"
            data-element="open-toast"
            onClick={() => setIsToastOpen(true)}
          >
            Show toast
          </Button>
        </Box>
      </Sidebar>
      <Toast
        open={isToastOpen}
        onDismiss={() => setIsToastOpen(false)}
        ref={toastRef}
        targetPortalId="stacked"
        data-role="toast"
      >
        Toast Message
      </Toast>
    </>
  );
};

export const TopModalOverride = () => {
  const [isOpen1, setIsOpen1] = useState(true);
  const [isOpen2, setIsOpen2] = useState(true);
  return (
    <>
      <Sidebar
        open={isOpen1}
        onCancel={() => setIsOpen1(false)}
        header="Sidebar"
        topModalOverride
      >
        foo
      </Sidebar>
      <Dialog open={isOpen2} onCancel={() => setIsOpen2(false)} title="Dialog">
        foo
      </Dialog>
    </>
  );
};
