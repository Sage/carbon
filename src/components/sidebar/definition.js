import { Sidebar } from './';
import DemoHelper from '../../utils/helpers/demo-helper';
import OptionsHelper from '../../utils/helpers/options-helper';
import podDefinition from '../pod/definition';

let definition = {
  component: Sidebar,
  key: 'sidebar',
  text: {
    bemClass: 'carbon-sidebar',
    details: '[content needed] Basic designs description for the component',
    description: '[content needed] Basic example of the component',
    name: 'Sidebar',
    type: 'modal'
  },
  defaultProps: Sidebar.defaultProps,
  props: Sidebar.propTypes,
  propOptions: {
    position: OptionsHelper.alignBinary(),
    size: OptionsHelper.sizesRestricted()
  }
};

definition.demoProps = {
  children: DemoHelper.elemArray(podDefinition, 7, 'title'),
  enableBackgroundUI: true,
  open: false,
  position: 'right',
  size: 'small'
};

export default definition;
