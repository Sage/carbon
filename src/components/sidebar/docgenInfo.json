{
  "src/components/sidebar/sidebar-header/sidebar-header.component.js": [
    {
      "description": "",
      "displayName": "SidebarHeader",
      "methods": [],
      "props": {
        "children": {
          "type": {
            "name": "node"
          },
          "required": false,
          "description": "This component supports children."
        },
        "className": {
          "type": {
            "name": "string"
          },
          "required": false,
          "description": "A custom class name."
        }
      }
    }
  ],
  "src/components/sidebar/sidebar.component.js": [
    {
      "description": "",
      "displayName": "Sidebar",
      "methods": [],
      "props": {
        "onCancel": {
          "type": {
            "name": "func"
          },
          "required": false,
          "description": ""
        },
        "open": {
          "type": {
            "name": "bool"
          },
          "required": false,
          "description": "A boolean to track the open state of the dialog"
        },
        "enableBackgroundUI": {
          "type": {
            "name": "bool"
          },
          "required": false,
          "description": "Set this prop to false to hide the translucent background when the dialog is open."
        },
        "position": {
          "type": {
            "name": "string"
          },
          "required": false,
          "description": "Sets the position of sidebar, either left or right."
        },
        "size": {
          "type": {
            "name": "string"
          },
          "required": false,
          "description": "Sets the size of the sidebar when open."
        }
      }
    }
  ]
}