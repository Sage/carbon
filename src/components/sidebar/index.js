export { default } from "./sidebar.component";
