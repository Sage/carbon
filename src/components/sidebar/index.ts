export { default } from "./sidebar.component";
export type { SidebarProps } from "./sidebar.component";
