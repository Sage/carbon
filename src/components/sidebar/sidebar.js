import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import SidebarHeader from './sidebar-header';
import Icon from '../icon';
import Modal from '../modal';

/**
 * A Sidebar widget.
 *
 * == How to use a Sidebar in a component:
 *
 * In your file
 *
 *   import { Sidebar } from 'carbon-react/lib/components/sidebar';
 *
 * To render the Sidebar:
 *
 *   <Sidebar
 *     onCancel={ closeSidebar }
 *     open={ true }
 *   />
 *
 * Side bar is positioned on the right hand screen of the window by default.
 * To position the sidebar on the left hand side pass `position='left'` to the component.
 *
 * The background behind the sidebar is disabled by default. To allow the user to interact
 * with all the UI pass `enableBackgroundUI={ true}` to the component
 *
 * @class Sidebar
 * @constructor
 */
class Sidebar extends Modal {
  static propTypes = {

    /**
     * A custom close event handler
     *
     * @property onCancel
     * @type {Function}
     */
    onCancel: PropTypes.func,

    /**
     * Sets the open state of the sidebar
     *
     * @property open
     * @type {Boolean}
     * @default false
     */
    open: PropTypes.bool,

    /**
     * Determines if the user can interact with
     * the background ui when the sidebar is open
     *
     * @property enableBackgroundUI
     * @type {Boolean}
     * @default false
     */
    enableBackgroundUI: PropTypes.bool,

    /**
     * Determines the position of the sidebar
     * 'left' or 'right'
     *
     * @property position
     * @type {String}
     * @default 'right'
     */
    position: PropTypes.string,

    /**
     * Size of dialog, default size is 450px
     *
     * @property size
     * @type {String}
     * @default medium
     */
    size: PropTypes.string
  }

  static defaultProps = {
    position: 'right',
    size: 'medium'
  }

  /**
   * Returns classes for the component.
   *
   * @method mainClasses
   * @return {String} Main className
   */
  get mainClasses() {
    return classNames(
      'carbon-sidebar',
      this.props.className
    );
  }

  /**
   * Returns classes for the sidebar.
   *
   * @method sidebarClasses
   * @return {String} sidebar className
   */
  get sidebarClasses() {
    return classNames(
      'carbon-sidebar__sidebar',
      `carbon-sidebar__sidebar--${this.props.position}`,
      `carbon-sidebar__sidebar--${this.props.size}`
    );
  }

  /**
   * Returns the markup for the close icon.
   *
   * @method closeButton
   * @return {Object} JSX
   */
  get closeButton() {
    if (this.props.onCancel) {
      return (
        <span className='carbon-sidebar__close'>
          <Icon
            className='carbon-sidebar__close-icon'
            data-element='close'
            onClick={ this.props.onCancel }
            type='close'
          />
        </span>
      );
    }
    return null;
  }

  componentTags(props) {
    return {
      'data-component': 'sidebar',
      'data-element': props['data-element'],
      'data-role': props['data-role']
    };
  }

  /**
   * Returns the computed HTML for the sidebar.
   *
   * @method sidebarHTML
   * @return {Object} JSX for sidebar
   */
  get modalHTML() {
    return (
      <div
        className={ this.sidebarClasses }
        { ...this.componentTags(this.props) }
      >
        { this.closeButton }
        { this.props.children }
      </div>
    );
  }

  get transitionName() {
    return `sidebar--${this.props.position}`;
  }
}

export {
  Sidebar,
  SidebarHeader
};
