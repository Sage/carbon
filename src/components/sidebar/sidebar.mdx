import { Meta, ArgTypes, Canvas } from "@storybook/blocks";
import TranslationKeysTable from "../../../.storybook/utils/translation-keys-table";

import * as SidebarStories from "./sidebar.stories.tsx";

<Meta title="Sidebar" of={SidebarStories} />

# Sidebar

<a
  target="_blank"
  href="https://zeroheight.com/2ccf2b601/p/925533-sidebar"
  style={{ color: "#007E45", fontWeight: "bold", textDecoration: "underline" }}
  rel="noreferrer"
>
  Product Design System component
</a>

## Contents

- [Quick Start](#quick-start)
- [Related Components](#related-components)
- [Examples](#examples)
- [Props](#props)

## Quick Start

```javascript
import Sidebar from "carbon-react/lib/components/sidebar";
```

- Sidebar is positioned on the right hand screen of the window by default. To position the sidebar on the left hand side pass `position='left'` to the component.

- The background behind the sidebar is disabled by default. To allow the user to interact with all the UI pass `enableBackgroundUI={ true }` to the component

## Related Components

- Simple task in context? [Try Dialog](https://carbon.sage.com/?path=/docs/dialog--default-story "Try Dialog").
- Complex task that needs more space? [Try Dialog Full Screen](https://carbon.sage.com/?path=/docs/dialog-full-screen--default "Try Dialog Full Screen").

## Examples

### Default

<Canvas of={SidebarStories.DefaultStory} />

### Custom padding around content

<Canvas of={SidebarStories.CustomPaddingAroundContent} />

### With header

<Canvas of={SidebarStories.WithHeader} />

### With scroll

If sidebar content does not fit the sidebar, scroll will appear. Default scroll theme is "light".

<Canvas of={SidebarStories.WithScroll} />

### With Typography

<Canvas of={SidebarStories.WithTypography} />

### Allowing other elements to be focused

Using the `focusableContainers` prop allows elements that are not DOM children of the Sidebar to be reachable with the keyboard while the Sidebar is open.
This may occasionally be useful with things like Toasts where they persist on the page until dismissed by clicking a close icon.
(In the example below, try clicking the cancel and/or save buttons and notice how you can navigate across one or both toasts as well as the sidebar using the keyboard.)

<Canvas of={SidebarStories.OtherFocusableContainers} />

### Custom width

It is possible to set a custom width for the `Sidebar` via the `width` [prop](#props).
Setting this prop will override the preset width value defined via the `size` prop.

<Canvas of={SidebarStories.CustomWidth} />

### With header and footer padding overridden

<Canvas of={SidebarStories.WithHeaderAndFooterPadding} />

### Top modal override

When multiple modals are open on a page Carbon manages the order internally so that the last one mounted in the DOM
is treated as the top one and the focus trap is set on it. However, this can potentially
create unexpected issues if the loading of the dialogs isn't synchronous. In order to guarantee that a given
modal is treated as top set the `topModalOverride` prop. Please note that in the scenario where this prop is
set on multiple modals, the last one mounted with it will be treated as top.

<Canvas of={SidebarStories.TopModalOverride} />

## Props

### Sidebar

<ArgTypes of={SidebarStories} />

## Translation keys

The following keys are available to override the translations for this component by passing in a custom locale object to the
[i18nProvider](../?path=/docs/documentation-i18n--docs).

<TranslationKeysTable
  translationData={[
    {
      name: "sidebar.ariaLabels.close",
      description: "The text for close button aria-label attribute",
      type: "func",
      returnType: "string",
    },
  ]}
/>
