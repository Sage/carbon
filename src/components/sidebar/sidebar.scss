$sidebar-animation-length: 250ms;

// sidebar right animation:
.sidebar--right-enter {
  opacity: 0;
  right: -30px;
}

.sidebar--right-enter.sidebar--right-enter-active {
  opacity: 1;
  right: 0;
  transition: all $sidebar-animation-length ease-in;
}

.sidebar--right-exit {
  right: 0;
  opacity: 1;
}

.sidebar--right-exit.sidebar--right-exit-active {
  opacity: 0;
  right: -50px;
  transition: all $sidebar-animation-length ease-in;
}

// sidebar left animation:
.sidebar--left-enter {
  opacity: 0;
  left: -30px;
}

.sidebar--left-enter.sidebar--left-enter-active {
  opacity: 1;
  left: 0;
  transition: all $sidebar-animation-length ease-in;
}

.sidebar--left-exit {
  left: 0;
  opacity: 1;
}

.sidebar--left-exit.sidebar--left-exit-active {
  opacity: 0;
  left: -50px;
  transition: all $sidebar-animation-length ease-in;
}
