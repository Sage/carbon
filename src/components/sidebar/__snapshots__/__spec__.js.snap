// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar sidebarClasses returns a base sidebar 1`] = `
<Portal
  key="1"
>
  <div
    className="carbon-sidebar custom-class"
    data-component="sidebar"
    data-element="bar"
    data-role="baz"
    data-state="open"
  >
    <CSSTransitionGroup
      component="div"
      transitionAppear={true}
      transitionAppearTimeout={500}
      transitionEnter={true}
      transitionEnterTimeout={500}
      transitionLeave={true}
      transitionLeaveTimeout={500}
      transitionName="modal-background"
    >
      <div
        className="carbon-modal__background"
      />
    </CSSTransitionGroup>
    <CSSTransitionGroup
      component="div"
      transitionAppear={true}
      transitionAppearTimeout={500}
      transitionEnter={true}
      transitionEnterTimeout={500}
      transitionLeave={true}
      transitionLeaveTimeout={500}
      transitionName="sidebar--right"
    >
      <div
        className="carbon-sidebar__sidebar carbon-sidebar__sidebar--right carbon-sidebar__sidebar--medium"
        data-component="sidebar"
        data-element="bar"
        data-role="baz"
      >
        <span
          className="carbon-sidebar__close"
        >
          <Icon
            bgSize="small"
            className="carbon-sidebar__close-icon"
            data-element="close"
            onClick={[Function]}
            type="close"
          />
        </span>
        <Textbox />
        <Textbox />
        <Textbox />
      </div>
    </CSSTransitionGroup>
  </div>
</Portal>
`;
